File: 1578-minimum-time-to-make-rope-colorful/1578-minimum-time-to-make-rope-colorful.js
/**
 * @param {string} colors
 * @param {number[]} neededTime
 * @return {number}
 */
var minCost = function(colors, neededTime) {
  let total = 0;
  let maxTime = 0;
  
  neededTime.forEach(function (time, i) {
    if (i > 0 && colors[i] !== colors[i - 1]) {
      maxTime = 0;
    }
    total += Math.min(maxTime, time);
    maxTime = Math.max(maxTime, time);
  });
  
  return total;
};

File: 1578-minimum-time-to-make-rope-colorful/1578-minimum-time-to-make-rope-colorful.go
func minCost(colors string, neededTime []int) (total int) {
    maxTime := neededTime[0]
    
    for i := 1; i < len(colors); i++ {
        if colors[i] != colors[i - 1] {
            maxTime = 0
        }
        total += min(maxTime, neededTime[i])
        maxTime = max(maxTime, neededTime[i])
    }
    
    return
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}